File: csharp/lc-1178/Solution.cs
using System;
using System.Collections.Generic;
using System.Numerics;

namespace lc_1178
{
    public class Solution
    {
        public IList<int> FindNumOfValidWords(string[] words, string[] puzzles)
        {
            var wordFreqDic = new Dictionary<uint, int>();
            foreach (var word in words)
            {
                var mask = 0u;
                foreach (var ch in word)
                {
                    mask |= (1u << (ch - 'a'));
                }

                if (BitOperations.PopCount(mask) <= 7)
                {
                    wordFreqDic[mask] = wordFreqDic.GetValueOrDefault(mask, 0) + 1;
                }
            }

            var ret = new List<int>();

            foreach (var puzzle in puzzles)
            {
                var mask = 0u;
                var maskFirstCh = 1u << (puzzle[0] - 'a');
                var puzzleSpan = puzzle.AsSpan();
                foreach (var ch in puzzleSpan.Slice(1))
                {
                    mask |= (1u << (ch - 'a'));
                }
                var numMatchingWords = 0;

                var subset = mask;
                do
                {
                    var wordPattern = subset | maskFirstCh;
                    numMatchingWords += wordFreqDic.GetValueOrDefault(wordPattern, 0);
                    subset = (subset - 1) & mask;
                } while (subset != mask);

                ret.Add(numMatchingWords);
            }

            return ret;
        }
    }
}

File: rust/lc-1178/src/solution.rs
pub struct Solution {}

impl Solution {
    pub fn find_num_of_valid_words(words: Vec<String>, puzzles: Vec<String>) -> Vec<i32> {
        let mut word_freq_map = std::collections::HashMap::new();
        for word in words {
            let mut mask = 0u32;
            for ch in word.bytes() {
                mask |= 1u32 << (ch - b'a');
            }

            if mask.count_ones() <= 7 {
                *word_freq_map.entry(mask).or_insert(0) += 1;
            }

        }

        let mut ret = Vec::with_capacity(puzzles.len());
        for puzzle in puzzles {
            let puzzle = puzzle.as_bytes();
            let mut mask = 0u32;
            let mask_first_ch = 1u32 << (puzzle[0] - b'a');

            for ch in puzzle[1..].iter() {
                mask |= 1u32 << (ch - b'a');
            }

            let mut num_matching_words = 0;
            let mut subset = mask;
            loop {
                let word_pattern = subset | mask_first_ch;
                num_matching_words += word_freq_map.get(&word_pattern).unwrap_or(&0);
                subset = subset.overflowing_sub(1u32).0 & mask;
                if subset == mask {
                    break;
                }
            }

            ret.push(num_matching_words);
        }

        ret
    }
}
